core/src/main/java/smile/projection/ICA.java

```diff
@@ -0,0 +1,365 @@
+    private int n;
+    /**
+     * constant to check if convergence on one component.
+     */
+    private double tolerance;
+    /**
+     * constant for maximum iteration for one component search.
+     */
+    private int maxIter;
+    /**
+     * The projected sample mean.
+     */
+    private double[] pmu;
+    /**
+     * Projection matrix.
+     */
+    private DenseMatrix projection;
+
+    /**
+     * determine the functions to calculate derivative and second-order derivative
+     */
+    private NegEntropyFunc funcMode = NegEntropyFunc.LOGCOSH;
+
+    /**
+     * functions to calculate derivative and second-order derivative
+     */
+    public enum NegEntropyFunc {
+        /**
+         * This Negative Entropy function is described as a "good general-purpose
+         * contrast function" in the original paper, and the default method used.
+         */
+        LOGCOSH,
+        /**
+         * This is function according to the paper may be better than {@link #LOGCOSH}
+         * when the independent components are highly super-Gaussian, or when robustness
+         * is very important
+         */
+        EXP
+    }
+
+    /**
+     * Constructor. Learn independent component analysis from the data. *
+     * 
+     * @param data
+     *            train data set
+     * @param p
+     *            target independent component count
+     */
+    public ICA(double[][] data, int p) {
+        this(data, p, 1E-5, 10000, NegEntropyFunc.LOGCOSH);
+    }
+
+    /**
+     * Constructor. Learn independent component analysis from the data.
+     * 
+     * @param data
+     *            train data set
+     * @param p
+     *            target independent component count
+     * @param tol
+     *            convergence threshold on component search
+     * @param maxIter
+     *            maximum iteration on component search
+     * @param negfunc
+     *            see {@link NegEntropyFunc}
+     */
+    public ICA(double[][] data, int p, double tol, int maxIter, NegEntropyFunc negfunc) {
+        int m = data.length;
+        this.n = data[0].length;
+        this.p = p;
+        this.tolerance = tol;
+        this.maxIter = maxIter;
+
+        projection = Matrix.zeros(p, n);
+
+        GaussianDistribution g = new GaussianDistribution(0, 1);
+        double[][] wps = new double[p][n];
+        for (int i = 0; i < p; i++) {
+            for (int j = 0; j < n; j++) {
+                wps[i][j] = g.rand();
+            }
+            // normalize
+            double norm = Math.norm2(wps[i]);
+            for (int j = 0; j < n; j++) {
+                wps[i][j] /= norm;
+            }
+        }
+
+        DenseMatrix whitened = whiteData(data);
+        for (int i = 0; i < p; i++) {
+            double[] lastwp = wps[i];
+
+            double diff = Double.MAX_VALUE;
+            int iterCnt = 0;
+            while (diff > this.tolerance) {
+                // step 1 calculate derivative of projection from non-linear functions
```

Comment:
This part is too complicated compared to the simple formula in the reference. We can simply it greatly if leverage Matrix interface more.